File: rust/src/protocol/v763/packets/play/clientbound/respawn.rs
//! Packet `respawn`. Hand-written.

use crate::codec::uuid_codec::{self as uuid_c, Uuid};
use crate::codec::{
    bitset, chat_component, identifier, nbt, position, slot,
    string_codec as string, varint, varlong, BytesReader, BytesWriter, Reader, Writer,
};
use crate::errors::ProtocolError;
use crate::protocol::v763::states::ConnectionState;
use crate::protocol::v763::ClientboundPacket;

pub const PACKET_ID: i32 = 0x41;

#[derive(Debug, Clone, PartialEq, Eq, Default)]
pub struct DeathLocation { pub dimension_name: String, pub location: (i32, i32, i32) }

#[derive(Debug, Clone, PartialEq, Default)]
pub struct Respawn {
    pub dimension: String,
    pub world_name: String,
    pub hashed_seed: i64,
    pub game_mode: u8,
    pub previous_game_mode: i8,
    pub is_debug: bool,
    pub is_flat: bool,
    pub copy_metadata: u8,
    pub death_location: Option<DeathLocation>,
    pub portal_cooldown: i32,
}

fn read_bool(reader: &mut BytesReader<'_>) -> Result<bool, ProtocolError> {
    let b = reader.read_exact(1)?[0];
    if b > 1 { return Err(ProtocolError::DecodeError(format!("bool: {}", b))); }
    Ok(b != 0)
}

impl Respawn {
    pub fn decode(reader: &mut BytesReader<'_>) -> Result<Self, ProtocolError> {
        let dimension = identifier::read(reader)?;
        let world_name = identifier::read(reader)?;
        let hb = reader.read_exact(8)?;
        let hashed_seed = i64::from_be_bytes([hb[0],hb[1],hb[2],hb[3],hb[4],hb[5],hb[6],hb[7]]);
        let game_mode = reader.read_exact(1)?[0];
        let previous_game_mode = reader.read_exact(1)?[0] as i8;
        let is_debug = read_bool(reader)?;
        let is_flat = read_bool(reader)?;
        let copy_metadata = reader.read_exact(1)?[0];
        let has_death = read_bool(reader)?;
        let death_location = if has_death {
            Some(DeathLocation {
                dimension_name: identifier::read(reader)?,
                location: position::read(reader)?,
            })
        } else { None };
        let portal_cooldown = varint::read(reader)?;
        Ok(Self { dimension, world_name, hashed_seed, game_mode, previous_game_mode,
                  is_debug, is_flat, copy_metadata, death_location, portal_cooldown })
    }
}

impl ClientboundPacket for Respawn {
    fn state(&self) -> ConnectionState { ConnectionState::Play }
    fn packet_id(&self) -> i32 { PACKET_ID }
    fn encode(&self, writer: &mut BytesWriter) -> Result<(), ProtocolError> {
        identifier::write(&self.dimension, writer)?;
        identifier::write(&self.world_name, writer)?;
        writer.write_all(&self.hashed_seed.to_be_bytes())?;
        writer.write_all(&[self.game_mode])?;
        writer.write_all(&[self.previous_game_mode as u8])?;
        writer.write_all(&[if self.is_debug { 1 } else { 0 }])?;
        writer.write_all(&[if self.is_flat { 1 } else { 0 }])?;
        writer.write_all(&[self.copy_metadata])?;
        match &self.death_location {
            None => writer.write_all(&[0])?,
            Some(d) => {
                writer.write_all(&[1])?;
                identifier::write(&d.dimension_name, writer)?;
                position::write(&d.location, writer)?;
            }
        }
        varint::write(self.portal_cooldown, writer)?;
        Ok(())
    }
}

File: python/minecraft_bot/protocol/v763/packets/play/clientbound/respawn.py
"""Packet `respawn` (play/clientbound, id 0x41).

Sent after a player dies and reuses their existing connection in a
(possibly different) dimension.
"""

from __future__ import annotations

import struct
from dataclasses import dataclass

from minecraft_bot.codec import Reader, Writer, position, string, varint
from minecraft_bot.errors import ValueOutOfRange

PACKET_ID = 0x41


@dataclass(frozen=True, slots=True)
class DeathLocation:
    dimension_name: str
    location: tuple[int, int, int]


@dataclass(frozen=True, slots=True)
class Respawn:
    dimension: str
    world_name: str
    hashed_seed: int            # i64
    gamemode: int               # i8
    previous_gamemode: int      # u8
    is_debug: bool
    is_flat: bool
    copy_metadata: bool
    death: DeathLocation | None
    portal_cooldown: int        # varint


def _read_bool(reader: Reader, field: str) -> bool:
    b = reader.read(1)[0]
    if b not in (0, 1):
        raise ValueOutOfRange(field, b)
    return b == 1


def decode(reader: Reader) -> Respawn:
    dim = string.read(reader)
    wn = string.read(reader)
    hs, = struct.unpack(">q", reader.read(8))
    gm, = struct.unpack(">b", reader.read(1))
    pgm = reader.read(1)[0]
    is_d = _read_bool(reader, "respawn.is_debug")
    is_f = _read_bool(reader, "respawn.is_flat")
    cm = _read_bool(reader, "respawn.copy_metadata")
    has_death = _read_bool(reader, "respawn.has_death")
    death = None
    if has_death:
        dn = string.read(reader)
        loc = position.read(reader)
        death = DeathLocation(dimension_name=dn, location=loc)
    pc = varint.read(reader)
    return Respawn(
        dimension=dim, world_name=wn, hashed_seed=hs,
        gamemode=gm, previous_gamemode=pgm,
        is_debug=is_d, is_flat=is_f, copy_metadata=cm,
        death=death, portal_cooldown=pc,
    )


def encode(packet: Respawn, writer: Writer) -> None:
    string.write(packet.dimension, writer)
    string.write(packet.world_name, writer)
    writer.write(struct.pack(">q", packet.hashed_seed))
    writer.write(struct.pack(">b", packet.gamemode))
    writer.write(bytes([packet.previous_gamemode & 0xFF]))
    for flag in (packet.is_debug, packet.is_flat, packet.copy_metadata):
        writer.write(b"\x01" if flag else b"\x00")
    if packet.death is None:
        writer.write(b"\x00")
    else:
        writer.write(b"\x01")
        string.write(packet.death.dimension_name, writer)
        position.write(packet.death.location, writer)
    varint.write(packet.portal_cooldown, writer)
